text/0000-embroider-v2-package-format.md

```diff
@@ -0,0 +1,772 @@
+
+It also needs to be possible for an addon published as V2 to work in existing apps on existing ember-cli versions. This is enabled by:
+
+- `ember-auto-import`, which serves as a high-fidelity polyfill for importing directly from NPM. V2 addons natively support importing from NPM, but they should depend on `ember-auto-import` so those imports will have the same meaning when used in classic ember-cli.
+- `@ember/macros`, which shall provide correct semantics regardless of whether the package using them is published as V1 or V2 and regardless of whether the build is being done by classic ember-cli or Embroider. Native V2 packages under Embroider can alway use macros, without an explicit dependency on `@ember/macros`, but they should include the dependency so that macros will work in classic ember-cli.
+- ember-cli already supports a `main` property in **Ember package metadata** and has supported it for many versions. This allows an addon to put its classic `index.js` file in a place other than the package's true `main`. This means that V2 addons can have their runtime `index.js` as `main`, and should point `ember-addon.main` to a `classic.js` file. The `classic.js` file should `require` and `export` a compatibility shim library that we will provide. The compatibility shim will have the classic methods like `treeForAddon`, `treeForPublic` that take the V2-formatted features and present them in a way that classic ember-cli will understand. Since V2 packages are much more static than V1 packages, this shim is expected to not be very complicated.
+
+# How we Teach This
+
+This RFC should have no direct impact on what app authors need to learn. They keep using addons the same way they always have. Future RFCs that take Embroider mainstream _will_ have impact, but that can be discussed then.
+
+The impact on addon authors is more significant. This design is fully backward compatible, and the intention is that all existing addons continue to work (some with worse compatibility hacks than others in the v1-to-v2 compiler). But there will be a demand for addons published in v2 format, since it is expected to result in faster build times. My prediction is that people who are motivated to get their own app build times down will send a lot of PRs to the addons they’re using.
+
+In many cases, converting addons to v2 makes them simpler. For example, today many addons use custom broccoli code to wrap third-party libraries in a fastboot guard that prevents the libraries from trying to load in Node (where they presumably don’t work). In v2, they can drop all that custom build-time code in favor of a macro-guarded `importSync`.
+
+This design does _not_ advocate loudly deprecating any v1 addon features. Doing that all at once would be unnecessarily disruptive. I would rather rely on the carrot of faster builds and Embroider stability than the stick of deprecation warnings. We can choose to deprecate v1 features in stages at a later time.
```

Comment:
Some additional points to cover in this section:

- Where to document this system (CLI guides)
- How to divide up the learning: general overveiw, change tutorial for new addons, write a tutorial for addon authors to help them migrate their addons, a tutorial for configuration that app users can reference if they run into problems with their addons, and abilities that unlock by having webpack compat. We would also need to revise some existing materials for regular ember-cli builds so that we appropriately present both options. (side note, what would we actually recommend as the happy path).